onadata/libs/utils/logger_tools.py

```diff
@@ -222,15 +223,22 @@ def create_instance(username, xml_file, media_files,
             username = username.lower()
 
         xml = xml_file.read()
+        xml_hash = Instance.get_hash(xml)
         xform = get_xform_from_submission(xml, username, uuid)
         check_submission_permissions(request, xform)
 
-        existing_instance_count = Instance.objects.filter(
-            xml=xml, xform__user=xform.user).count()
-
-        if existing_instance_count > 0:
-            existing_instance = Instance.objects.filter(
-                xml=xml, xform__user=xform.user)[0]
+        # Check if an instance whose content hashes to the same value is already in the DB.
+        duplicate_instances_exist = Instance.objects.filter(xml_hash=xml_hash).exists()
+        # Also check against unhashed `Instance`s if no matching hash was found.
+        duplicate_instances_exist = duplicate_instances_exist or Instance.objects.filter(
+            xml_hash=Instance.DEFAULT_XML_HASH, xml=xml, xform__user=xform.user).exists()
+
+        if duplicate_instances_exist:
+            duplicate_instances = Instance.objects.filter(xml_hash=xml_hash)
+            if not duplicate_instances.exists():
+                duplicate_instances = Instance.objects.filter(
+                    xml_hash=Instance.DEFAULT_XML_HASH, xml=xml, xform__user=xform.user)
+            existing_instance = duplicate_instances[0]
```

Comment:
May as well `try` getting the first element and handle the `IndexError` if it gets thrown. That should avoid separate DB queries for checking existence and object retrieval